python/lsst/dbb/buffmngrs/endpoint/ingester/commands.py

```diff
@@ -0,0 +1,120 @@
+from .ingester import Ingester
+from .. import validation
+from ..utils import setup_logger
+
+
+logger = logging.getLogger(__name__)
+
+
+@click.group()
+def ingester():
+    pass
+
+
+@ingester.command()
+@click.option("-v", "--validate", is_flag=True, default=False,
+              help="Validate configuration before starting the service.")
+@click.argument("filename", type=click.Path(exists=True))
+def start(filename, validate):
+    """Starts an ingester using a configuration from FILENAME.
+    """
+    with open(filename) as f:
+        configuration = yaml.safe_load(f)
+    if validate:
+        schema = yaml.safe_load(validation.ingester)
+        try:
+            jsonschema.validate(instance=configuration, schema=schema)
+        except jsonschema.ValidationError as ex:
+            raise ValueError(f"Configuration error: {ex.message}.")
+        except jsonschema.SchemaError as ex:
+            raise ValueError(f"Schema error: {ex.message}.")
+        return
+
+    config = configuration.get("logging", None)
+    setup_logger(logging.getLogger(), options=config)
+
+    logger.info("setting up database connection...")
+    config = configuration["database"]
+
+    module = importlib.import_module("sqlalchemy.pool")
+    pool_name = config.get("pool_class", "QueuePool")
+    try:
+        class_ = getattr(module, pool_name)
+    except AttributeError:
+        msg = f"unknown connection pool type: {pool_name}"
+        logger.error(msg)
+        raise RuntimeError(msg)
+
+    engine = create_engine(config["engine"],
+                           echo=config.get("echo", False),
+                           poolclass=class_)
+
+    logger.info("checking if required database tables exist...")
+    required = {table for table in config["tablenames"].values()}
+    available = set(inspect(engine).get_table_names())
+    missing = required - available
+    if missing:
+        msg = f"table(s) {', '.join(missing)} not found in the database"
+        logger.error(msg)
+        raise RuntimeError(msg)
+
+    Session = sessionmaker(bind=engine)
+    session = Session()
+
+    mapper = config["tablenames"]
+
+    logger.info("setting up Ingester...")
+    config = configuration["ingester"]
+    config["session"] = session
+    config["orms"] = mapper
+
+    # Configure ingest plugin.
+    package_name = "lsst.dbb.buffmngrs.endpoint.ingester"
+    module = importlib.import_module(".plugins", package=package_name)
+    plugin_name = config["plugin"]["name"]
+    plugin_config = config["plugin"]["config"]
+    try:
+        class_ = getattr(module, plugin_name)
+    except AttributeError as ex:
+        msg = f"Unknown ingest plugin '{plugin_name}'."
+        logger.error(msg)
+        raise RuntimeError(msg)
+    else:
+        try:
+            plugin = class_(plugin_config)
+        except ValueError as ex:
+            msg = f"{class_.__name__}: invalid configuration: {ex}."
+            raise RuntimeError(msg)
+        config["plugin"] = plugin
```

Comment:
(Assuming I read this code right and ingester object is being instantiated here) with init methods creating objects like a Gen3 butler, there are no problems making a single object with database connections here and share it among threaded workers later?  (BTW took me a while to track this down because in the later thread code I thought config["plugin"] was still the dictionary from the config file.)

Reply:
I see your point. Indeed I was "covertly" adding, overwriting existing settings in the `finder/ingester` section of the global configuration. I'll rewrite relevant parts to make it clear what is the configuration read from the file and what is the actual configuration for the Finder/Ingester.